version: '3.3'

services:
  db-node-1:
    build:
      context: mysql
      dockerfile: Dockerfile
    container_name: db-node-1
    restart: unless-stopped
    environment:
      MYSQL_DATABASE: archdb
      MYSQL_USER: login
      MYSQL_PASSWORD: password
      MYSQL_ROOT_PASSWORD: '1'
    command: >
      --sql-mode='STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
      --general-log=ON
      --log-queries-not-using-indexes=ON
    networks:
      - arch-network
    ports:
      - 3360:3306
    volumes:
      - db-node-1-data:/var/lib/mysql

  db-node-2:
    build:
      context: mysql
      dockerfile: Dockerfile
    container_name: db-node-2
    restart: unless-stopped
    environment:
      MYSQL_DATABASE: archdb
      MYSQL_USER: login
      MYSQL_PASSWORD: password
      MYSQL_ROOT_PASSWORD: '1'
    command: >
      --sql-mode='STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
      --general-log=ON
      --log-queries-not-using-indexes=ON
    networks:
      - arch-network
    ports:
      - 3361:3306
    volumes:
      - db-node-2-data:/var/lib/mysql

  db-node-3:
    build:
      context: mysql
      dockerfile: Dockerfile
    container_name: db-node-3
    restart: unless-stopped
    environment:
      MYSQL_DATABASE: archdb
      MYSQL_USER: login
      MYSQL_PASSWORD: password
      MYSQL_ROOT_PASSWORD: '1'
    command: >
      --sql-mode='STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION'
      --general-log=ON
      --log-queries-not-using-indexes=ON
    networks:
      - arch-network
    ports:
      - 3362:3306
    volumes:
      - db-node-3-data:/var/lib/mysql

  proxysql:
    build:
      context: proxysql
      dockerfile: Dockerfile
    container_name: proxysql
    command: >
      --initial
    volumes:
      - proxysql-data:/var/lib/proxysql
    healthcheck:
      test: "/usr/bin/mysql --user=login --password=password --execute \"SHOW DATABASES;\""
      timeout: 20s
      retries: 5
    depends_on:
      - db-node-1
      - db-node-2
      - db-node-3

    networks:
      - arch-network
    ports:
      # Mysql Client Port
      - "6033:6033"
      # Mysql Admin Port
      - "6032:6032"

  cache:
    image: redis:6.2-alpine
    restart: always
    ports:
      - '6379:6379'
    command: redis-server --save 20 1 --loglevel warning
    networks:
      - arch-network
    volumes:
      - cache:/data

  social_network:
    build:
      context: social_network
      dockerfile: Dockerfile
    container_name: social_network
    environment:
      - DATABASE_HOST=proxysql
      - DATABASE_PORT=6033
      - DATABASE_LOGIN=login
      - DATABASE_PASSWORD=password
      - DATABASE_DATABASE=archdb
      - CACHE=cache:6379
    ports:
      - 8082:8080
    expose:
      - "8082"
    networks:
      - arch-network
    depends_on:
      - proxysql
      - cache
      - db-node-1
      - db-node-2
      - db-node-3

volumes:
  db-node-1-data:
  db-node-2-data:
  db-node-3-data:
  proxysql-data:
  cache:

networks:
  arch-network:
    driver: "bridge"
    external: false
